### PR TITLE
attest API support, software based ECDH Certificate and related changes.

### PR DESCRIPTION
**Pull request by srinibas-zededa**
Signed-off-by: Srinibas Maharana <srinibas@zededa.com>
Please review. 
1. attest API on zedagent, for publishing the device certificates
2. rephrase cert API changes for both MITM and cipher object
3. changes in zedagent and ciper module for Ecdh Cert changes for related modules
4. Software ECDH Certificate create and signing with device key in tpmmgr
5. Publishing them to the zedagent, for controller consumption

Will be adding device certificate based signing for soft ECDH Certificate soon.

### CODE REVIEW

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go [hunk 1]
```diff
@@ -851,6 +878,7 @@ func createEcdhCert() error {
 		if err != nil {
 			return err
 		}
+		defer rw.Close()
```
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go:881]: Good catch.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go [hunk 2]
```diff
@@ -199,19 +325,105 @@ func sendCertsProtobuf(attestReq *attest.ZAttestReq, iteration int) {
...
+	// start the controller certificate fetch task
+	go controllerCertsTask(ctx, ctx.cipherCtx.triggerControllerCerts)
+}
+
+// Controller certificate, check whether there is a Sha mismatch
+// to trigger the post request
+func handleControllerCertsSha(ctx *zedagentContext,
+	config *zconfig.EdgeDevConfig) {
+	sumHash := hex.EncodeToString(ctx.cipherCtx.cfgControllerCertHash)
+	certHash := config.GetControllercertConfighash()
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlecertconfig.go:366]: If this is not set by the controller (I think that means you get an empty string), then we shouldn't trigger because that just wastes bandwidth.
Instead you can log that the ControllercertConfigHash is not set.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go [hunk 3]
```diff
@@ -199,19 +325,105 @@ func sendCertsProtobuf(attestReq *attest.ZAttestReq, iteration int) {
...
+		triggerControllerCertEvent(ctx)
+	}
+}
+
+// Eve node certificate, check whether there is a Sha mismatch
+// to trigger the post request
+func handleEveNodeCertsSha(ctx *zedagentContext,
+	config *zconfig.EdgeDevConfig) {
+	sumHash := hex.EncodeToString(ctx.cipherCtx.eveNodeCertHash)
+	certHash := config.GetNodecertConfighash()
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlecertconfig.go:377]: Ditto, since the controller doesn't yet implement this.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go [hunk 4]
```diff
@@ -199,19 +325,105 @@ func sendCertsProtobuf(attestReq *attest.ZAttestReq, iteration int) {
...
+
+//  eve node certificate post trigger function
+func triggerEveNodeCertEvent(ctxPtr *zedagentContext) {
+
+	log.Info("Trigger Eve Node Certs publish")
+	select {
+	case ctxPtr.cipherCtx.triggerEveNodeCerts <- struct{}{}:
+		// Do nothing more
+	default:
+		log.Warnf("Failed to send on triggerEveNodeCerts")
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlecertconfig.go:403]: Don't know if we have this for other cases following the same pattern, but it would be helpful to a new reader to add a comment saying this means that the trigger has already been done and not yet processed, hence no need to re-trigger.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go [hunk 5]
```diff
@@ -199,19 +325,105 @@ func sendCertsProtobuf(attestReq *attest.ZAttestReq, iteration int) {
...
+	}
+}
+
+func convertLocalToApiHashAlgo(algo types.CertHashType) evecommon.HashAlgorithm {
+	switch algo {
+	case types.CertHashTypeSha256First16:
+		return evecommon.HashAlgorithm_HASH_ALGORITHM_SHA256_16BYTES
+	default:
+		return evecommon.HashAlgorithm_HASH_ALGORITHM_INVALID
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlecertconfig.go:412]: Since this is local to API it would be more helpful if this was a log.Fatal so when a new hash is added it is more obvious that a case is missing. The log.Fatal should include the %d of the valiue of the algo.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go [hunk 6]
```diff
@@ -199,19 +325,105 @@ func sendCertsProtobuf(attestReq *attest.ZAttestReq, iteration int) {
...
+	case types.CertTypeOnboarding:
+		return zcert.ZCertType_CERT_TYPE_DEVICE_ONBOARDING
+	case types.CertTypeRestrictSigning:
+		return zcert.ZCertType_CERT_TYPE_DEVICE_RESTRICTED_SIGNING
+	case types.CertTypeEk:
+		return zcert.ZCertType_CERT_TYPE_DEVICE_ENDORSEMENT_RSA
+	case types.CertTypeEcdhXchange:
+		return zcert.ZCertType_CERT_TYPE_DEVICE_ECDH_EXCHANGE
+	default:
+		return zcert.ZCertType_CERT_TYPE_CONTROLLER_NONE
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlecertconfig.go:427]: Since this is local to API it would be more helpful if this was a log.Fatal so when a new cert type is added it is more obvious that a case is missing. The log.Fatal should include the %d of the value of the certType.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go [hunk 7]
```diff
@@ -127,70 +142,181 @@ func unpublishControllerCert(ctx *getconfigContext, key string) {
...
+func sendEveNodeCertsProtobuf(attestReq *attest.ZAttestReq, iteration int) {
 	data, err := proto.Marshal(attestReq)
 	if err != nil {
 		log.Fatal("SendInfoProtobufStr proto marshaling error: ", err)
 	}
 
-	deviceUUID := zcdevUUID.String()
-	zedcloud.RemoveDeferred("attest:" + deviceUUID)
+	reqID := zcdevUUID.String()
+	zedcloud.RemoveDeferred("attest:" + reqID)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlecertconfig.go:319]: I would be more maintainable if you defined a local deferKey := "attest:"+reqID and use it here plus in the SetDeferred below.
Non-obvious things would happen if the two ever get out of synch due to some future code changes.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go [hunk 8]
```diff
@@ -127,70 +142,181 @@ func unpublishControllerCert(ctx *getconfigContext, key string) {
...
+	certURL := zedcloud.URLPathString(serverNameAndPort,
+		zedcloudCtx.V2API, nilUUID, "certs")
+
+	// not V2API
+	if !zedcloud.UseV2API() {
+		return false
+	}
+
+	reqID := zcdevUUID.String()
+	zedcloud.RemoveDeferred("ccerts:" + reqID)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlecertconfig.go:197]: I would be more maintainable if you defined a local deferKey := "ccerts:"+reqID and use it here plus in the SetDeferred below.
Non-obvious things would happen if the two ever get out of synch due to some future code changes.

But WAIT, this is a get and not a POST, right? So there is no deferred processing needed. Just log the error and presumably a trigger will happen again.
When does the hash used by the trigger get updated? For it to trigger again you should wait to update the hash of the controller certs until you have received the certs - I haven't checked if that is already the case in this code.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go [hunk 9]
```diff
@@ -127,70 +142,181 @@ func unpublishControllerCert(ctx *getconfigContext, key string) {
...
+		}
+		zedcloud.SetDeferred("ccerts:"+reqID, nil, 0, certURL, zedcloudCtx, true)
+		return false
+	}
+
+	switch resp.StatusCode {
+	case http.StatusOK, http.StatusCreated, http.StatusNotModified:
+		log.Infof("getCertsFromController: status %s", resp.Status)
+	case http.StatusNotFound, http.StatusUnauthorized, http.StatusNotImplemented, http.StatusBadRequest:
+		log.Infof("getCertsFromController: server %s does not support V2 API", serverName)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlecertconfig.go:215]: I think this info message is potentially misleading. Could get StatusNotFound etc for other reasons?
Would it be better to just have the default case printing of the status code?

File: pkg/pillar/cmd/zedagent/handlecertconfig.go (hunk 8 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedagent/handlecertconfig.go:197]: Checked and that is indeed the case. So you can remove the RemoveDeferred and SetDeferred for this get.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go [hunk 10]
```diff
@@ -127,70 +142,181 @@ func unpublishControllerCert(ctx *getconfigContext, key string) {
...
+	triggerEveNodeCertEvent(ctx)
 	return
 }
 
 // Run a task certificate post task, on change trigger
-func certsTask(ctx *zedagentContext, triggerCerts <-chan struct{}) {
-	iteration := 0
+func controllerCertsTask(ctx *zedagentContext, triggerCerts <-chan struct{}) {
+
 	log.Infoln("starting certs publish task")
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlecertconfig.go:165]: Comment and log are off; this is a fetch not a publish task.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go [hunk 11]
```diff
@@ -127,70 +142,181 @@ func unpublishControllerCert(ctx *getconfigContext, key string) {
...
-	publishCerts(ctx, iteration)
+	getCertsFromController(ctx)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlecertconfig.go:166]: I guess we need this now when we do not have a hash sent from the controller. But once the controller sends the hash we could use the trigger (plus saving the controller certs in a file) to avoid any fetches until there is a change. But no need to change it now.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go (hunk 3 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedagent/handlecertconfig.go:377]: Actually it is better to simplify this by removing the ctx.cipherCtx.eveNodeCertHash. pubsub will only call the handleModify when there is a change which means you need to trigger sending a message.

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go [hunk 12]
```diff
@@ -1104,8 +1239,9 @@ func Run(ps *pubsub.PubSub) {
 
 		pubEveNodeCert, err := ps.NewPublication(
 			pubsub.PublicationOptions{
-				AgentName: agentName,
-				TopicType: types.EveNodeCert{},
+				AgentName:  agentName,
+				TopicType:  types.EveNodeCert{},
+				Persistent: true,
```
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go:1244]: Why do you need this to be persistent? I shouldn't hurt but I wonder if there is some subtle ordering issue which requires it.

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go [hunk 13]
```diff
@@ -908,6 +936,109 @@ func createEcdhCert() error {
...
+		NotAfter:  notAfter,
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+	return template
+}
+
+// get the file based software ECDH private key
```
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go:966]: From the comment, name, and code I can't tell what his function does.
Code and comment seems to imply a get, and if the get succeeds it generates a key.

I think part of the naming issue is that getEveNodePrivateKey is really getting the debice private key.
We call that a deviceKey and not an EveNode key. 
Plus this comment should say "generate the software ECDH key and certificate, where the certificate is signed using the device privaye key".

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go [hunk 14]
```diff
@@ -908,6 +936,109 @@ func createEcdhCert() error {
 	return nil
 }
 
+// create Ecdh Template
+func createEcdhSoftTemplate() x509.Certificate {
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, _ := rand.Int(rand.Reader, serialNumberLimit)
+	notBefore := time.Now()
+	// twenty years
+	notAfter := notBefore.AddDate(20, 0, 0)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go:945]: Do these lifetimes match what we use for the TPM-based ECDH cert?

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go [hunk 15]
```diff
@@ -908,6 +936,109 @@ func createEcdhCert() error {
...
+
+	template := x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			Country:      []string{"US"},
+			Province:     []string{"CA"},
+			Locality:     []string{"Santa Clara"},
+			Organization: []string{"Zededa, Inc"},
+			CommonName:   "Cipher Block",
```
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go:954]: That common name does not match what we use for ECDH on existing devices. 
I suggest we call both of them the same and have it be "Device ECDH certificate"

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go [hunk 16]
```diff
@@ -796,9 +806,13 @@ func testEcdhAES() error {
 	return nil
 }
 
-func getDevicePrivateKey() (*ecdsa.PrivateKey, error) {
-	// XXX:TBD, currently only one private key
-	keyPEMBlock, err := ioutil.ReadFile(types.DeviceKeyName)
+// software private key
+func getEveNodePrivateKey(isDeviceKey bool) (*ecdsa.PrivateKey, error) {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go:810]: The old name is correct; the new name is very confusing.

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go [hunk 17]
```diff
@@ -719,12 +725,16 @@ func DecryptSecretWithEcdhKey(X, Y *big.Int, iv, ciphertext, plaintext []byte) e
...
-	// when TPM is not enabled, use the locally stored private key
-	if !etpm.IsTpmEnabled() {
-		privateKey, err := getDevicePrivateKey()
+func getDecryptKey(X, Y *big.Int, eveNodeCert *types.EveNodeCert) ([32]byte, error) {
+	// check whether this is a software certificate
+	if !etpm.IsTpmEnabled() || !eveNodeCert.IsTpm {
+		isDeviceKey := false
+		if !etpm.IsTpmEnabled() && eveNodeCert == nil {
+			isDeviceKey = true
+		}
```
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go:734]: Is this to handle old devices? Wouldn't they get a ECDH certificate created on boot when they run this new code?
In any case, if this code needs to stay it needs comments explaining why one would ever use the deviceKey for ECDH.

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go (hunk 16 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/tpmmgr/tpmmgr.go:810]: If you want a readPrivateKey(filename string) to avoid duplicating code that's fine. getDevicePrivateKey and getDeviceECDHPrivateKey can call that function.

File: pkg/pillar/cipher/handlecipher.go [hunk 18]
```diff
@@ -169,18 +155,22 @@ func DecryptCipherBlock(ctx *DecryptCipherContext,
 }
 
 func decryptCipherBlockWithECDH(ctx *DecryptCipherContext,
-	cipherBlock types.CipherBlockStatus) ([]byte, error) {
-	cert, err := getControllerCertEcdhKey(ctx, cipherBlock)
+	cipherContext *types.CipherContext, cipherBlock types.CipherBlockStatus) ([]byte, error) {
+	ecdhKey, err := getControllerCertEcdhKey(ctx, cipherContext.ControllerCertKey())
```
Comment (eriknordmark) [inline pkg/pillar/cipher/handlecipher.go:159]: This function returns a certificate, right? If so why did you rename the local variable to ecdhKey?

File: pkg/pillar/cipher/handlecipher.go [hunk 19]
```diff
@@ -169,18 +155,22 @@ func DecryptCipherBlock(ctx *DecryptCipherContext,
...
-	cipherContext := lookupCipherContext(ctx, cipherBlock.CipherContextID)
-	if cipherContext == nil {
-		errStr := fmt.Sprintf("cipher context %s not found\n",
-			cipherBlock.CipherContextID)
+	eveNodeCert := lookupEveNodeCert(ctx, cipherContext.EveNodeCertKey())
+	if eveNodeCert == nil {
+		log.Errorf("Eve Node Certificate get fail")
+		return []byte{}, err
+	}
+	if eveNodeCert.HasError() {
```
Comment (eriknordmark) [inline pkg/pillar/cipher/handlecipher.go:169]: When would you have errors on the certificate?

File: pkg/pillar/cipher/handlecipher.go [hunk 20]
```diff
@@ -203,15 +193,22 @@ func decryptCipherBlockWithECDH(ctx *DecryptCipherContext,
...
+func getControllerCertEcdhKey(ctx *DecryptCipherContext, key string) (*ecdsa.PublicKey, error) {
+	config := lookupControllerCert(ctx, key)
+	if config == nil {
+		errStr := fmt.Sprintf("Controller Certificate get fail")
+		log.Error(errStr)
+		return nil, errors.New(errStr)
+	}
+	if config.HasError() {
+		errStr := fmt.Sprintf("controller certificate has error: %v",
+			config.Error)
```
Comment (eriknordmark) [inline pkg/pillar/cipher/handlecipher.go:205]: When would you have an error?

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go [hunk 21]
```diff
@@ -79,6 +80,9 @@ const (
 	//location of the ecdh certificate
 	ecdhCertFile = types.IdentityDirname + "/ecdh.cert.pem"
 
+	//location of the ecdh private key
```
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go:83]: Please add "on devices without a TPM"

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go [hunk 22]
```diff
@@ -630,6 +634,7 @@ func printPCRs() {
 		fmt.Printf("Error opening TPM device: %s", err)
 		return
 	}
+	defer rw.Close()
```
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go:637]: Good catch

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go [hunk 23]
```diff
@@ -787,7 +797,7 @@ func testEcdhAES() error {
 	aesEncrypt(ciphertext, msg, encryptKey[:], iv)
 
 	recoveredMsg := make([]byte, len(ciphertext))
-	err = DecryptSecretWithEcdhKey(publicAX, publicAY, iv, ciphertext, recoveredMsg)
+	err = DecryptSecretWithEcdhKey(publicAX, publicAY, nil, iv, ciphertext, recoveredMsg)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go:800]: What does it mean to pass nil here? (I don't know how this test function is used).
A comment explaing why nil and what it does would be helpful

File: pkg/pillar/cmd/zedagent/handlecertconfig.go (hunk 2 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/handlecertconfig.go:366]: sure, will do.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go (hunk 3 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/handlecertconfig.go:377]: sure

File: pkg/pillar/cmd/zedagent/handleconfig.go [hunk 24]
```diff
@@ -216,8 +215,8 @@ func getLatestConfig(url string, iteration int,
 				getconfigCtx.configGetStatus = types.ConfigGetTemporaryFail
 			}
 		} else if rtf == types.SenderStatusCertMiss {
-			// trigger a one sec timer to acquire new certs from cloud
-			triggerFetchCerts(getconfigCtx.zedagentCtx)
+			// trigger to acquire new controller certs from cloud
+			triggerControllerCertEvent(getconfigCtx.zedagentCtx)
```
Comment (vikas-zededa) [inline pkg/pillar/cmd/zedagent/handleconfig.go:219]: As per https://wiki.lfedge.org/display/EVE/Measured+Boot+and+Remote+Attestation, we should be triggering attestation when config fails with 403. Perhaps I am not able to find it but is that already handled? Per my cursory reading of the code, seems like we return SenderStatusNone for anything other than http.StatusOK, http.StatusCreated and http.StatusNotModified?

Comment (eriknordmark) [reply pkg/pillar/cmd/zedagent/handleconfig.go:219]: > As per https://wiki.lfedge.org/display/EVE/Measured+Boot+and+Remote+Attestation, we should be triggering attestation when config fails with 403. Perhaps I am not able to find it but is that already handled? Per my cursory reading of the code, seems like we return SenderStatusNone for anything other than http.StatusOK, http.StatusCreated and http.StatusNotModified?

We haven't added support for 403 yet. That requires some new flow where zedagent tells tpmmgr to re-attest, which isn't in place yet.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go (hunk 4 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/handlecertconfig.go:392]: sure

File: pkg/pillar/cmd/zedagent/handlecertconfig.go (hunk 5 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/handlecertconfig.go:401]: ok

File: pkg/pillar/cmd/zedagent/handlecertconfig.go (hunk 7 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/handlecertconfig.go]: sure

File: pkg/pillar/cmd/zedagent/handlecertconfig.go (hunk 8 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/handlecertconfig.go]: removed.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go (hunk 10 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/handlecertconfig.go:164]: sure

File: pkg/pillar/cmd/zedagent/handlecertconfig.go (hunk 11 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/handlecertconfig.go:165]: ok

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go (hunk 12 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/tpmmgr/tpmmgr.go]: removed

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go (hunk 17 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/tpmmgr/tpmmgr.go:738]: Added the comments.

File: pkg/pillar/cipher/handlecipher.go (hunk 19 above)
Comment (srinibas-zededa) [reply pkg/pillar/cipher/handlecipher.go:169]: For validation/siging errors, if any.

File: pkg/pillar/cipher/handlecipher.go (hunk 18 above)
Comment (srinibas-zededa) [reply pkg/pillar/cipher/handlecipher.go:159]: reverted.

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go (hunk 23 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/tpmmgr/tpmmgr.go:804]: added the comment.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go (hunk 9 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/handlecertconfig.go:211]: sure done.

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go (hunk 12 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/tpmmgr/tpmmgr.go]: Did you check that WiFi decryption doesn't depend on it first? I honestly don't know exactly what it depends on to decrypt the WiFi credentials.

File: pkg/pillar/cipher/handlecipher.go (hunk 19 above)
Comment (eriknordmark) [reply pkg/pillar/cipher/handlecipher.go:169]: But I don't see any code in this PR setting an error. Am I missing it?

Comment (srinibas-zededa) [reply pkg/pillar/cipher/handlecipher.go:169]: The controller certificate validation, is pending for object cipher. This is already there for MITM, will add that functionality.
